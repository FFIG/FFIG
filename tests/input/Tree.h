#include <memory>
#include <random>

#ifdef __clang__
#define C_API __attribute__((annotate("GENERATE_C_API")))
#else
#define C_API
#endif

static std::mt19937 mt;
static std::uniform_int_distribution<int> d(1,10);
static auto gen = []{return d(mt);};

class C_API Tree
{
  int data_;
  std::unique_ptr<Tree> left_;
  std::unique_ptr<Tree> right_;

  public:

  Tree(int levels=0)
  {
    data_ = gen();
    if ( levels <= 0 ) return;
    left_ = std::make_unique<Tree>(levels-1);
    right_ = std::make_unique<Tree>(levels-1);
  }

  const Tree* left_subtree() const
  {
    return left_.get();
  }

  const Tree* right_subtree() const
  {
    return right_.get();
  }

  void erase_data() 
  {
    data_ = 0;
  }

  int data() const
  {
    return data_;
  }
  
  void set_data(int x)
  {
    data_ = x;
  }

};

