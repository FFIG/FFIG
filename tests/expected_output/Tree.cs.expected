// This code was generated by FFIG <http://ffig.org>.
// Manual edits will be lost.
using System;
using System.Runtime.InteropServices;

namespace Tree_c {

  public class Exception : System.Exception {

    [DllImport("Tree_c")]
    private static extern void Tree_clear_error();

    [DllImport("Tree_c")]
    private static extern IntPtr Tree_error();

    public Exception() : base(Marshal.PtrToStringAnsi(Tree_error())) {
      Tree_clear_error();
    }
  }

  public class Tree {

    [DllImport("Tree_c")]
    private static extern int Tree_Tree_dispose(IntPtr c_obj);

    [DllImport("Tree_c")]
    private static extern int Tree_Tree_create(int levels, out IntPtr ptr);
    [DllImport("Tree_c")]
    private static extern IntPtr Tree_Tree_left_subtree_noexcept(IntPtr c_obj);
    [DllImport("Tree_c")]
    private static extern IntPtr Tree_Tree_right_subtree_noexcept(IntPtr c_obj);
    [DllImport("Tree_c")]
    private static extern int Tree_Tree_data_noexcept(IntPtr c_obj);
    [DllImport("Tree_c")]
    private static extern void Tree_Tree_set_data_noexcept(IntPtr c_obj, int x);

    protected IntPtr c_obj_;

    public Tree(int levels) {
      int rc = Tree_Tree_create(levels, out c_obj_);
      if(rc != 0) {
        throw new Tree_c.Exception();
      }
    }

    protected Tree() {
    }

    protected Tree(IntPtr c_obj) {
      c_obj_ = c_obj;
    }

    ~Tree() {
      Tree_Tree_dispose(c_obj_);
    }
    public Tree left_subtree() {
      var rv = Tree_Tree_left_subtree_noexcept(c_obj_);
      if(rv == IntPtr.Zero) {
        return null;
      }
      return new Tree(rv);
    }
    public Tree right_subtree() {
      var rv = Tree_Tree_right_subtree_noexcept(c_obj_);
      if(rv == IntPtr.Zero) {
        return null;
      }
      return new Tree(rv);
    }
    public int data() {
      var rv = Tree_Tree_data_noexcept(c_obj_);
      return rv;
    }
    public void set_data(int x) {
      Tree_Tree_set_data_noexcept(c_obj_, x);
    }
  }

}
